### PR TITLE
refactor and avoid needless probe

### PR DESCRIPTION
**Pull request by jp-bennett**

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -557,9 +557,28 @@ void setup()
     screen = new graphics::Screen(screen_found, screen_model, screen_geometry);
 
     readFromRTC(); // read the main CPU RTC at first (in case we can't get GPS time)
-    gps = createGps();
-    if (gps)
+    int8_t rx_gpio = config.position.rx_gpio;
```
[the remark this answers is not in the document]
Comment (jp-bennett) [reply src/main.cpp:560]: I was trying to make the code a bit easier to follow, but I'll admit I'm not used to using the factory paradigm.

File: variants/heltec_v3/variant.h [hunk 2]
```diff
@@ -11,7 +11,9 @@
 #define VEXT_ENABLE Vext // active low, powers the oled display and the lora antenna boost
 #define BUTTON_PIN 0
 
-#define PIN_GPS_EN 46 // GPS power enable pin
+// #define GPS_TX_PIN UINT32_MAX
```
[the remark this answers is not in the document]
Comment (jp-bennett) [reply variants/heltec_v3/variant.h:14]: Good catch

File: src/gps/GPS.cpp [hunk 3]
```diff
@@ -14,8 +14,8 @@
 #endif
 
 // If we have a serial GPS port it will not be null
-#ifdef GPS_SERIAL_NUM
-HardwareSerial _serial_gps_real(GPS_SERIAL_NUM);
+#ifdef ARCH_ESP32
+HardwareSerial _serial_gps_real(1);
```
[the remark this answers is not in the document]
Comment (jp-bennett) [reply src/gps/GPS.cpp:18]: Seems to work!

File: variants/rpipico/variant.h [hunk 4]
```diff
@@ -13,7 +13,6 @@
 
 #define NO_GPS 1
```
[the remark this answers is not in the document]
Comment (jp-bennett) [reply variants/rpipico/variant.h:14]: Not used, out it goes

File: variants/tbeam/variant.h [hunk 5]
```diff
@@ -37,4 +37,6 @@
 // and waking from light sleep
 // #define PMU_IRQ 35
 #define HAS_AXP192
-#define GPS_UBLOX
\ No newline at end of file
+#define GPS_UBLOX
+#define GPS_RX_PIN 34
```
[the remark this answers is not in the document]
Comment (jp-bennett) [reply variants/tbeam/variant.h:41]: tbeam0.7 variant.h already has different pins defined. But this is the sort of thing most likely to be missed.

File: src/main.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jp-bennett) [reply src/main.cpp:560]: I'm seeing a paradigm called a "named constructor" that seems perfect for this. Refactoring to that.

[the remark this answers is not in the document]
Comment (mverch67) [reply src/main.cpp:560]: In any case such pattern decouple the classes and make the code more readable and maintainable.

File: src/sleep.h [hunk 6]
```diff
@@ -19,7 +19,7 @@ extern XPowersLibInterface *PMU;
 #endif
 
 void setGPSPower(bool on);
-void doGPSpowersave(bool on);
+
```
Comment (thebentern) [inline src/sleep.h:22]: Huge win!

File: src/gps/GPS.cpp [hunk 7]
```diff
@@ -576,6 +567,8 @@ int32_t GPS::runOnce()
             return 0;
         }
     }
+    if (config.position.gps_enabled == false)
+        return 0;
```
Comment (thebentern) [inline src/gps/GPS.cpp:571]: Can we `return disable();` here, or do we leave the thread open for toggling?

Comment (jp-bennett) [reply src/gps/GPS.cpp:571]: Correct, this is so the triple-click will work. I'm not opposed to adding a re-enable function to go with the disable(), but that seems out of scope for this patch set.

**Review by thebentern — APPROVED** [review]

**Pull request merged**